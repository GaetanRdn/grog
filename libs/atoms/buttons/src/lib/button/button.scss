@mixin gro-button($config) {
  $typography: map-get($config, typography);
  $colors: map-get($config, colors);

  .gro-button {
    appearance: none;
    padding: 0.25rem 0.75rem;
  }

  @each $size in small, medium, large {
    $typo: map-get($typography, button-#{$size});
    $font-size: map-get($typo, font-size);

    .gro-#{$size} {
      font-size: $font-size;
      font-family: map-get($typo, font-family);
      font-weight: map-get($typo, font-weight);
      border-radius: 0.25 * $font-size;
    }
  }

  @each $color in primary, accent, warn {
    $color-config: map-get($colors, $color);
    $light-config: map-get($colors, $color + '-light');
    $dark-config: map-get($colors, $color + '-dark');

    .gro-#{$color} {
      &:not(.gro-outlined) {
        color: map-get($color-config, color);
        background-color: map-get($color-config, background-color);
        border: none;

        &:hover {
          color: map-get($dark-config, color);
          background-color: map-get($dark-config, background-color);
        }
      }

      &.gro-outlined {
        color: map-get($color-config, background-color);
        background-color: map-get($color-config, color);
        border: 1px solid map-get($color-config, border-color);

        &:hover {
          color: map-get($light-config, color);
          background-color: map-get($light-config, background-color);
        }
      }
    }
  }
}
