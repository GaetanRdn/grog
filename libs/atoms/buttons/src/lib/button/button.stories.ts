import { CommonModule } from '@angular/common';
import { Meta, Story } from '@storybook/angular';
import { ButtonDirective } from './button.directive';

export default {
  title: 'atoms/button',
  component: ButtonDirective,
  parameters: {
    jest: ['button.directive.spec.ts'],
  },
} as Meta<ButtonDirective>;

const Template: Story<ButtonDirective> = (args: ButtonDirective) => ({
  props: args,
  moduleMetadata: { declarations: [ButtonDirective], imports: [CommonModule] },
  template: `<button groButton [size]="size" [outlined]="outlined" [color]="color">Click</button>`,
});

export const Default = Template.bind({});
