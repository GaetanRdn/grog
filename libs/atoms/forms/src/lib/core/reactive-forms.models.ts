export type OnChangeFn<ValueType> = (value: ValueType | null) => void;
export type OnTouchedFn = () => void;
