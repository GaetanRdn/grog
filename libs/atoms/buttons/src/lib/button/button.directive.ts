import { Directive, Input, NgModule } from '@angular/core';
import { CoerceBoolean } from '@grorg/decorators';

@Directive({
  selector: 'button[groButton]',
  // standalone: true,
  host: {
    class: 'gro-button',
    '[class.gro-small]': 'size === "small"',
    '[class.gro-medium]': 'size === "medium"',
    '[class.gro-large]': 'size === "large"',
    '[class.gro-outlined]': 'outlined',
    '[class.gro-primary]': 'color === "primary"',
    '[class.gro-accent]': 'color === "accent"',
    '[class.gro-warn]': 'color === "warn"',
  },
})
export class ButtonDirective {
  @Input()
  public size: 'small' | 'medium' | 'large' = 'medium';

  @Input()
  @CoerceBoolean()
  public outlined?: boolean;

  @Input()
  public color: 'primary' | 'accent' | 'warn' = 'primary';
}

@NgModule({
  declarations: [ButtonDirective],
  exports: [ButtonDirective],
})
export class ButtonModule {}
