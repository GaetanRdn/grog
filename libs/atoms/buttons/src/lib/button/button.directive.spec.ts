import { Component } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ButtonDirective } from './button.directive';

describe('ButtonDirective', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ButtonDirective, DefaultComponent, WithSizeComponent, CoerceOutlinedComponent, WithColorsComponent]
    });
  });

  describe('Default', () => {
    let fixture: ComponentFixture<DefaultComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(DefaultComponent);
    });

    test('should create default button', () => {
      // WHEN
      fixture.detectChanges();

      // THEN
      expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
    });
  });

  describe('Size cases', () => {
    let fixture: ComponentFixture<WithSizeComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(WithSizeComponent);
    });

    test.each([
      'small',
      'medium',
      'large'
    ] as ('small' | 'medium' | 'large')[])('when size is %s then check snapshot', (size: 'small' | 'medium' | 'large') => {
      // GIVEN
      fixture.componentInstance.size = size;

      // WHEN
      fixture.detectChanges();

      // THEN
      expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
    });
  });

  describe('Outlined cases', () => {
    let fixture: ComponentFixture<CoerceOutlinedComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(CoerceOutlinedComponent);
    });

    test('should create', () => {
      // WHEN
      fixture.detectChanges();

      // THEN
      expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
    });

  });

  describe('Color cases', () => {
    let fixture: ComponentFixture<WithColorsComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(WithColorsComponent);
    });

    test.each([
      'primary',
      'accent',
      'warn'
    ] as ('primary' | 'accent' | 'warn')[])('when color is %s then check snapshot', (color: 'primary' | 'accent' | 'warn') => {
      // GIVEN
      fixture.componentInstance.color = color;

      // WHEN
      fixture.detectChanges();

      // THEN
      expect(fixture.debugElement.children[0].nativeElement).toMatchSnapshot();
    });
  });
});

@Component({
  template: `
      <button groButton>Click Me</button>`
})
class DefaultComponent {
}

@Component({
  template: `
      <button groButton size="{{size}}">Click Me</button>`
})
class WithSizeComponent {
  public size!: 'small' | 'medium' | 'large';
}

@Component({
  template: `
      <button groButton color="{{color}}">Click Me</button>`
})
class WithColorsComponent {
  public color?: 'primary' | 'accent' | 'warn';
}

@Component({
  template: `
      <button groButton outlined>Click Me</button>`
})
class CoerceOutlinedComponent {
}
