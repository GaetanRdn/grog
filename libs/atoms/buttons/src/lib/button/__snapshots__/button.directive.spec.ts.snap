// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonDirective Color cases when color is accent then check snapshot 1`] = `
<button
  class="gro-button gro-medium gro-accent"
  grobutton=""
>
  Click Me
</button>
`;

exports[`ButtonDirective Color cases when color is primary then check snapshot 1`] = `
<button
  class="gro-button gro-medium gro-primary"
  grobutton=""
>
  Click Me
</button>
`;

exports[`ButtonDirective Color cases when color is warn then check snapshot 1`] = `
<button
  class="gro-button gro-medium gro-warn"
  grobutton=""
>
  Click Me
</button>
`;

exports[`ButtonDirective Default should create default button 1`] = `
<button
  class="gro-button gro-medium gro-primary"
  grobutton=""
>
  Click Me
</button>
`;

exports[`ButtonDirective Outlined cases should create 1`] = `
<button
  class="gro-button gro-medium gro-outlined gro-primary"
  grobutton=""
  outlined=""
>
  Click Me
</button>
`;

exports[`ButtonDirective Size cases when size is large then check snapshot 1`] = `
<button
  class="gro-button gro-large gro-primary"
  grobutton=""
>
  Click Me
</button>
`;

exports[`ButtonDirective Size cases when size is medium then check snapshot 1`] = `
<button
  class="gro-button gro-medium gro-primary"
  grobutton=""
>
  Click Me
</button>
`;

exports[`ButtonDirective Size cases when size is small then check snapshot 1`] = `
<button
  class="gro-button gro-small gro-primary"
  grobutton=""
>
  Click Me
</button>
`;
